### PR TITLE
[ENH] replace apply linalg norm with native pandas ufuncs

### LINKED ISSUE #591: [ENH] Performance: `_energy_default` on BaseDistribution is very slow
**Is your feature request related to a problem? Please describe.**

Using `.apply(np.linalg.norm, axis=1, ord=1)` causes an between a 3x and 5x increase in total runtime when compared to using a native pandas implementation of the L1 Norm.

The portion of the code itself has a 3000x increase in runtime (in my use case - 20,000 data points in a time series). This is masked by some other inefficient pandas usage (I will raise this also).

In a Jupyter notebook:

```python
import numpy as np
import pandas as pd

y = pd.DataFrame(np.random.randn(20_000, 1), columns=["target"])
```
```python
%timeit y.apply(np.linalg.norm, axis=1, ord=1)
```
> 98.4 ms ± 1.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

```python
%timeit y.abs().sum(axis=1)
```
> 710 μs ± 3.04 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

As we are applying `np.linalg.norm` across `axis=1` this should be equivalent to taking the L1 vector norm of each row. This in turn is simply the sum of the abs values which is trivially vectorised by pandas.

```python
a = y.apply(np.linalg.norm, axis=1, ord=1)
b = y.abs().sum(axis=1)

all(a == b)
```
> True

**Describe the solution you'd like**
Change [line L1207](https://github.com/sktime/skpro/blob/main/skpro/distributions/base/_base.py#L1207)
from:
```python
        energy = spl.apply(np.linalg.norm, axis=1, ord=1)
```

to

```python
        energy = spl.abs().sum(axis=1)
```

for a 140x speed-up (and 3000x speed-up on my dataset)

**Describe alternatives you've considered**
None. The solution is very simple and give excellent speed-ups.

**Additional context**
I am not 100% clear on how edge cases such as +/-inf, NaN etc are handled in `np.linalg.norm`. Pandas is generally pretty flexible with `skipna` and `min_count` kwargs. Equally, vectorised checks with `.where` or `.mask` can be easily added at minimal cost.

### PR DESCRIPTION
**Pull request by joshdunnlime**
#### Reference Issues/PRs
Fixes #591

#### What does this implement/fix? Explain your changes.
Very simply, it replaces a pandas `.apply` (which under the hood is just a loop) with a pandas/numpy native ufunc. This mean that is operation is now fully vectorised.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?

Edge cases such as how the reviewer expects values such as `np.inf` and `np.NaN` to be handled.

#### Did you add any tests for the change?

No tests added. All previous tests passed.

#### Any other comments?
Reduction in total run time for line `_base.py:1207` was from 16.56 seconds to 0.11 seconds (150x speed-up).

Modified example from `examples/01_skpro_intro.ipynb`:

```python
import pandas as pd

from sklearn.datasets import load_diabetes
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold

from skpro.benchmarking.evaluate import evaluate
from skpro.metrics import CRPS
from skpro.regression.residual import ResidualDouble

# 1. specify dataset]
# create more data to show vectorisation on larger dataset
n_copies = 10
X, y = load_diabetes(return_X_y=True, as_frame=True)
X = pd.concat([X] * n_copies, ignore_index=True)
y = pd.concat([y] * n_copies, ignore_index=True)

# 2. specify estimator
estimator = ResidualDouble(LinearRegression(), distr_type="t")

# 3. specify cross-validation schema
cv = KFold(n_splits=3)

# 4. specify evaluation metric
crps = CRPS()

# 5. evaluate - run the benchmark
results = evaluate(estimator=estimator, X=X, y=y, cv=cv, scoring=crps)
```

Left: Run with linalg.norm. Right: Run with .abs().sum(axis=1).

Run with linalg.norm

Run with .abs().sum(axis=1)

Profiler displays are with [Flamegraph](https://marketplace.visualstudio.com/items?itemName=rafaelha.vscode-flamegraph) in VS Code.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/skpro/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/skpro/blob/main/.all-contributorsrc) in the `skpro` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by joshdunnlime** [comment]
Basically, if `df.apply()` is being used there is a good chance this is not as performant as it could be. The exception here is if it is applying a `np.xxx` vectorised function or a lambda that operates at the pandas level (`lambda s: s.mean` for example).

I did a quick search of the code base and found no obvious issues.

There was another slightly niche but large performance issue with the use of `to_numpy` (which is recommended in pandas) with tz-aware datetimes on the distribution base class. The repeated calls of `to_numpy` on a tz-aware datetime index gave HUGE slow down (100,000x vs utc or non-tz-aware DatetimeIndex). This is a pandas -> numpy issue where numpy doesn't support timezone-aware datetimes. I raised an issue with pandas [here](https://github.com/pandas-dev/pandas/issues/62309#issuecomment-3294093815) but, as numpy is unlikely to add tz-aware datetimes, pandas cannot improve this. They might consider raising a warning. As mentioned, this issue is magnified in skpro due to the repeated calls of `to_numpy` in `_get_bc_params`. It might be worth while skpro raising a performance warning if tz-aware DatetimeIndex is used (or any index of non-numpy objects). Happy to raise this as a separate issue.

**Comment by fkiraly** [comment]
> Happy to raise this as a separate issue.

That would be great, thanks!

A solution would be to move `_get_bc_params` closer to `numpy`?

**Comment by fkiraly** [comment]
PS: I also see you are working on the statmixedml repos - there are ongoing discussions about integration with `skpro`.

Would you be interested to have a quick coordination session, I would suggest @tingiskhan (who added zero-inflated nad hurdle distributions, and is working on `prophetverse` and `sktime` too), @statmixedml, and you?
On the `sktime` discord https://discord.com/invite/54ACzaFsn7

**Comment by joshdunnlime** [comment]
> > Happy to raise this as a separate issue.
> 
> That would be great, thanks!
> 
> A solution would be to move `_get_bc_params` closer to `numpy`?

Issue raised here: https://github.com/sktime/skpro/issues/597
